### PR TITLE
fix(Settings): write settings to disk when they change

### PR DESCRIPTION
**Pull request by vith**
This prevents losing settings changes when qTox is terminated abnormally,
such as by power loss or crash.

issue #1969

I have another branch in progress with unix signal handlers to quit gracefully when the system is shutting down as well.

---

This change is [Reviewable](https://reviewable.io/reviews/qtox/qtox/3192)

### CODE REVIEW

File: src/persistence/settingstransaction.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+ * The motivation behind this design is to easily prevent cascading
+ * changes from triggering multiple writes to disk. The destructor
+ * function only makes a save call if the object being destroyed
+ * was the last instance of its class alive.
+ *
+ * Think: "last one out closes the door."
+ */
+SettingsTransaction::SettingsTransaction()
+{
+//    qDebug() << "created" << this;
```
Comment (sudden6) [inline src/persistence/settingstransaction.cpp]: Commented out Debug statements shouldn't be in the final code. Just remove them with another commit so they are visible in the git log.

**Comment by vith** [comment]
Done.

**Comment by sudden6** [comment]
This looks ok at the first glance, but will probably take some time to review thoroughly

**Comment by vith** [comment]
Had some discussion on IRC. Todo:
- [ ] Add a throttle/debounce timer so SettingsTransaction can be used on things like window resize events.

**Comment by zetok** [comment]
#2733 ?

**Comment by zetok** [comment]
Hi,

As you might be aware, repo has been moved under `qTox` org on github.

What does it have to do with your PR, you might ask.

Well, it doesn't change all that much, except that workflow has been slightly improved – now there are required checks that PR needs to pass in order to be merged.

Those checks are:
- travis (checks whether PR builds & commit message format)
- reviewable (requirement for PR to be reviewed & accepted)

Since osx travis was broken for a bit, you will need to rebase your PR on newest master to get it to pass travis check. Sorry for inconvenience.

Reminder: as always, PR needs to be possible to merge without conflicts. Make sure it does ;)

**Comment by iphydf** [comment]
@vith are you going to rebase this branch?

**Comment by zetok** [comment]
PR didn't have any activity for over 2 months, and no one seems to be interested in working on it, thus closing.

If that's not the case, please either ask to reopen the pull, or make a new one.

**Pull request closed without merging**